api/grpc/mpi/v1/file_stream.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package v1
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"google.golang.org/grpc"
+)
+
+var (
+	ErrInvalidHeader     = errors.New("invalid header")
+	ErrUnexpectedContent = errors.New("unexpected content")
+	ErrSend              = errors.New("send error")
+	ErrWrite             = errors.New("write error")
+	ErrFailedRead        = errors.New("failed to read")
+)
+
+// SendChunkedFile reads the src into [FileDataChunkContent]s, and sends a valid sequence of
+// [FileDataChunk]s down the stream.
+func SendChunkedFile(
+	meta *MessageMeta,
+	header FileDataChunk_Header,
+	src io.Reader,
+	dst grpc.ServerStreamingServer[FileDataChunk],
+) error {
+	chunkCount := int(header.Header.GetChunks())
+	chunkSize := int(header.Header.GetChunkSize())
+	total := int(header.Header.GetFileMeta().GetSize())
+	if chunkSize == 0 || chunkCount == 0 || total == 0 {
+		return fmt.Errorf("%w:  %v", ErrInvalidHeader, header.Header)
+	}
+	if err := dst.Send(&FileDataChunk{
+		Meta:  meta,
+		Chunk: &header,
+	}); err != nil {
+		return fmt.Errorf("%w: %s (header)", ErrSend, err)
```

Comment:
nit: the `%s` might be losing details about the error that might be relevant for callers. We can retain it using [`errors.Join`](https://pkg.go.dev/errors#Join)

```suggestion
		return errors.Join(fmt.Errorf("%w: header", ErrSend), err)
```

Or we make a bigger change to define an error type that supports [`unwrap`](https://pkg.go.dev/errors#Unwrap)